File: csharp/GreekPostalCodeDb/PostalCodeClient.cs
using System.Globalization;
using System.Reflection;
using System.Text;
using System.Text.RegularExpressions;
using Microsoft.Data.Sqlite;

namespace GreekPostalCodeDb;

/// <summary>An offline, read-only client for the bundled Greek postal-code database.</summary>
public sealed class PostalCodeClient : IDisposable
{
    private static readonly Regex PostcodePattern = new("^\\d{5}$", RegexOptions.Compiled | RegexOptions.CultureInvariant);
    private static readonly Regex RangeNumberPattern = new("^\\s*(\\d+)", RegexOptions.Compiled | RegexOptions.CultureInvariant);
    private SqliteConnection? _connection;
    private readonly string _databasePath;

    private PostalCodeClient(string databasePath, SqliteConnection connection) => (_databasePath, _connection) = (databasePath, connection);

    /// <summary>Opens the database embedded in this package in read-only mode.</summary>
    public static PostalCodeClient Create()
    {
        var path = Path.Combine(Path.GetTempPath(), $"greek-postal-code-db-{Guid.NewGuid():N}.sqlite");
        try
        {
            using var source = typeof(PostalCodeClient).Assembly.GetManifestResourceStream("GreekPostalCodeDb.library.sqlite")
                ?? throw new InvalidOperationException("The bundled library.sqlite resource is missing.");
            using (var target = File.Create(path)) source.CopyTo(target);
            var connection = new SqliteConnection(new SqliteConnectionStringBuilder { DataSource = path, Mode = SqliteOpenMode.ReadOnly }.ToString());
            connection.Open();
            using var command = connection.CreateCommand();
            command.CommandText = "PRAGMA query_only = ON;";
            command.ExecuteNonQuery();
            return new PostalCodeClient(path, connection);
        }
        catch
        {
            File.Delete(path);
            throw;
        }
    }

    public void Dispose()
    {
        _connection?.Dispose();
        _connection = null;
        try { File.Delete(_databasePath); } catch (IOException) { }
        GC.SuppressFinalize(this);
    }

    public IReadOnlyList<Entity> ListRegions(ListOptions? options = null) => List("regions", null, null, options ?? new(), false);
    public IReadOnlyList<Entity> ListRegionalUnits(RegionalUnitOptions? options = null) => List("regional_units", "region_id", options?.RegionId, options ?? new(), options?.IncludeOfficialCode ?? false);
    public IReadOnlyList<Entity> ListMunicipalities(MunicipalityOptions? options = null) => List("municipalities", "regional_unit_id", options?.RegionalUnitId, options ?? new(), options?.IncludeOfficialCode ?? false);
    public IReadOnlyList<Entity> ListMunicipalUnits(MunicipalityChildOptions? options = null) => List("municipal_units", "municipality_id", options?.MunicipalityId, options ?? new(), options?.IncludeOfficialCode ?? false);
    public IReadOnlyList<Entity> ListCommunities(MunicipalityChildOptions? options = null) => List("communities", "municipality_id", options?.MunicipalityId, options ?? new(), options?.IncludeOfficialCode ?? false);

    public IReadOnlyList<Entity> SearchRegions(string query, ListOptions? options = null) => Search("regions", null, query, null, options ?? new(), false);
    public IReadOnlyList<Entity> SearchRegionalUnits(string query, RegionalUnitOptions? options = null) => Search("regional_units", "region_id", query, options?.RegionId, options ?? new(), options?.IncludeOfficialCode ?? false);
    public IReadOnlyList<Entity> SearchMunicipalities(string query, MunicipalityOptions? options = null) => Search("municipalities", "regional_unit_id", query, options?.RegionalUnitId, options ?? new(), options?.IncludeOfficialCode ?? false);
    public IReadOnlyList<Entity> SearchMunicipalUnits(string query, MunicipalityChildOptions? options = null) => Search("municipal_units", "municipality_id", query, options?.MunicipalityId, options ?? new(), options?.IncludeOfficialCode ?? false);
    public IReadOnlyList<Entity> SearchCommunities(string query, MunicipalityChildOptions? options = null) => Search("communities", "municipality_id", query, options?.MunicipalityId, options ?? new(), options?.IncludeOfficialCode ?? false);

    public PostcodeResult? GetPostcode(string postcode, PostcodeOptions? options = null)
    {
        if (!PostcodePattern.IsMatch(postcode)) return null;
        options ??= new();
        var result = One("SELECT postcode, latitude, longitude, local_area, municipal_unit_id, community_id, municipality_id FROM locations WHERE postcode = $postcode", command => command.Parameters.AddWithValue("$postcode", postcode), Location);
        if (result is null) return null;
        return result with
        {
            Hierarchy = options.IncludeHierarchy ? PostcodeHierarchy(result) : null,
            Streets = options.IncludeStreets ? Streets(postcode) : null,
        };
    }

    public AddressValidation ValidateAddress(AddressInput input)
    {
        var postcode = GetPostcode(input.Postcode, new(true, input.Street is not null || input.HouseNumber is not null));
        var postcodeResult = !PostcodePattern.IsMatch(input.Postcode)
            ? Invalid(input.Postcode, "postcode_must_be_exactly_five_digits")
            : postcode is null ? Invalid(input.Postcode, "postcode_not_found") : Valid(input.Postcode, [postcode]);
        if (postcode is null)
            return new(postcodeResult,
                input.Street is null ? null : NotEvaluated(input.Street, "postcode_not_found"),
                input.HouseNumber is null ? null : NotEvaluated(input.HouseNumber, "postcode_not_found"),
                input.Municipality is null ? null : NotEvaluated(input.Municipality, "postcode_not_found"),
                input.MunicipalUnit is null ? null : NotEvaluated(input.MunicipalUnit, "postcode_not_found"),
                input.Community is null ? null : NotEvaluated(input.Community, "postcode_not_found"),
                input.RegionalUnit is null ? null : NotEvaluated(input.RegionalUnit, "postcode_not_found"),
                input.Region is null ? null : NotEvaluated(input.Region, "postcode_not_found"));

        var hierarchy = postcode.Hierarchy!;
        var street = input.Street is null ? null : ValidateStreet(input.Street, postcode.Streets!);
        var houseNumber = input.HouseNumber is null ? null : ValidateHouseNumber(input.HouseNumber, street);
        return new(postcodeResult, street, houseNumber,
            input.Municipality is null ? null : ValidateReference("municipalities", input.Municipality, hierarchy.Municipality),
            input.MunicipalUnit is null ? null : ValidateReference("municipal_units", input.MunicipalUnit, hierarchy.MunicipalUnit),
            input.Community is null ? null : ValidateReference("communities", input.Community, hierarchy.Community),
            input.RegionalUnit is null ? null : ValidateReference("regional_units", input.RegionalUnit, hierarchy.RegionalUnit),
            input.Region is null ? null : ValidateReference("regions", input.Region, hierarchy.Region));
    }

    private IReadOnlyList<Entity> List(string table, string? parentColumn, long? parentId, ListOptions options, bool officialCode)
    {
        ValidateLimit(options.Limit);
        var sql = $"SELECT id, name{(officialCode ? ", official_code" : "")} FROM {table}" + (parentColumn is not null && parentId is not null ? $" WHERE {parentColumn} = $parent" : "") + " ORDER BY name, id";
        var entities = All(sql, command => { if (parentColumn is not null && parentId is not null) command.Parameters.AddWithValue("$parent", parentId.Value); }, row => Entity(row, officialCode));
        var withHierarchy = options.IncludeHierarchy ? entities.Select(entity => entity with { Hierarchy = EntityHierarchy(table, entity.Id) }) : entities;
        return options.Limit is int limit ? withHierarchy.Take(limit).ToArray() : withHierarchy.ToArray();
    }

    private IReadOnlyList<Entity> Search(string table, string? parentColumn, string query, long? parentId, ListOptions options, bool officialCode)
    {
        var normalized = Normalize(query);
        if (normalized.Length == 0) return [];
        var matches = List(table, parentColumn, parentId, options with { IncludeHierarchy = false, Limit = null }, officialCode)
            .Where(entity => Normalize(entity.Name).StartsWith(normalized, StringComparison.Ordinal))
            .Select(entity => options.IncludeHierarchy ? entity with { Hierarchy = EntityHierarchy(table, entity.Id) } : entity)
            .ToArray();
        return options.Limit is int limit ? matches.Take(limit).ToArray() : matches;
    }

    private EntityHierarchy EntityHierarchy(string table, long id) => table switch
    {
        "regions" => new(DecentralizedAdministration: Named(One("SELECT da.id, da.name FROM regions r JOIN decentralized_administrations da ON da.id = r.decentralized_administration_id WHERE r.id = $id", c => c.Parameters.AddWithValue("$id", id), row => Entity(row, false)))),
        "regional_units" => RegionalUnitHierarchy(id),
        "municipalities" => MunicipalityHierarchy(id),
        _ => MunicipalityChildHierarchy(ScalarLong($"SELECT municipality_id FROM {table} WHERE id = $id", id)),
    };

    private EntityHierarchy PostcodeHierarchy(PostcodeResult location)
    {
        var municipalUnit = location.MunicipalUnitId is long unitId ? Coded(One("SELECT id, name, official_code FROM municipal_units WHERE id = $id", c => c.Parameters.AddWithValue("$id", unitId), row => Entity(row, true))) : null;
        var community = location.CommunityId is long communityId ? Coded(One("SELECT id, name, official_code FROM communities WHERE id = $id", c => c.Parameters.AddWithValue("$id", communityId), row => Entity(row, true))) : null;
        var municipality = location.MunicipalityId is long municipalityId ? Coded(One("SELECT id, name, official_code FROM municipalities WHERE id = $id", c => c.Parameters.AddWithValue("$id", municipalityId), row => Entity(row, true))) : null;
        var parent = municipality is null ? new EntityHierarchy() : MunicipalityHierarchy(municipality.Id);
        return parent with { Municipality = municipality, MunicipalUnit = municipalUnit, Community = community };
    }

    private EntityHierarchy RegionalUnitHierarchy(long id)
    {
        var region = Named(One("SELECT r.id, r.name FROM regional_units ru JOIN regions r ON r.id = ru.region_id WHERE ru.id = $id", c => c.Parameters.AddWithValue("$id", id), row => Entity(row, false)));
        var administration = Named(One("SELECT da.id, da.name FROM regional_units ru JOIN regions r ON r.id = ru.region_id JOIN decentralized_administrations da ON da.id = r.decentralized_administration_id WHERE ru.id = $id", c => c.Parameters.AddWithValue("$id", id), row => Entity(row, false)));
        return new(DecentralizedAdministration: administration, Region: region);
    }

    private EntityHierarchy MunicipalityHierarchy(long id)
    {
        var unit = Coded(One("SELECT ru.id, ru.name, ru.official_code FROM municipalities m JOIN regional_units ru ON ru.id = m.regional_unit_id WHERE m.id = $id", c => c.Parameters.AddWithValue("$id", id), row => Entity(row, true)));
        var parent = unit is null ? new EntityHierarchy() : RegionalUnitHierarchy(unit.Id);
        return parent with { RegionalUnit = unit };
    }

    private EntityHierarchy MunicipalityChildHierarchy(long? municipalityId)
    {
        if (municipalityId is null) return new();
        var municipality = Coded(One("SELECT id, name, official_code FROM municipalities WHERE id = $id", c => c.Parameters.AddWithValue("$id", municipalityId.Value), row => Entity(row, true)));
        var parent = municipality is null ? new EntityHierarchy() : MunicipalityHierarchy(municipality.Id);
        return parent with { Municipality = municipality };
    }

    private IReadOnlyList<Street> Streets(string postcode) => All("SELECT id, postcode, name, odd_start, odd_end, even_start, even_end FROM streets WHERE postcode = $postcode ORDER BY name, id", c => c.Parameters.AddWithValue("$postcode", postcode), Street);

    private ValidationResult ValidateReference(string table, LocalityReference reference, Entity? linked)
    {
        if (reference.Id is null && string.IsNullOrWhiteSpace(reference.Name)) return Invalid(reference, "reference_must_not_be_empty", []);
        var officialCode = table != "regions";
        var matches = reference.Id is long id
            ? All($"SELECT id, name{(officialCode ? ", official_code" : "")} FROM {table} WHERE id = $id", c => c.Parameters.AddWithValue("$id", id), row => Entity(row, officialCode))
            : List(table, null, null, new(), officialCode).Where(entity => Normalize(entity.Name).StartsWith(Normalize(reference.Name!), StringComparison.Ordinal)).ToArray();
        var values = matches.Cast<object>().ToArray();
        if (linked is null) return Invalid(reference, "postcode_has_no_linked_entity", values);
        return matches.Any(candidate => candidate.Id == linked.Id) ? Valid(reference, values) : Invalid(reference, null, values);
    }

    private static ValidationResult ValidateStreet(string street, IReadOnlyList<Street> streets)
    {
        var matches = streets.Where(candidate => Normalize(candidate.Name) == Normalize(street)).Cast<object>().ToArray();
        return matches.Length > 0 ? Valid(street, matches) : Invalid(street, "street_not_found_for_postcode", matches);
    }

    private static ValidationResult ValidateHouseNumber(object value, ValidationResult? street)
    {
        if (street?.Status != ValidationStatus.Valid) return NotEvaluated(value, "street_is_required_and_must_be_valid");
        if (!PositiveInteger(value, out var number)) return Invalid(value, "house_number_must_be_a_positive_integer");
        var streets = street.Matches!.Cast<Street>().ToArray();
        var checks = streets.Select(candidate => ContainsHouseNumber(candidate, number)).Where(check => check is not null).Cast<bool>().ToArray();
        return checks.Length == 0 || checks.Any(check => check) ? Valid(value, streets.Cast<object>().ToArray(), checks.Length == 0 ? "street_has_no_usable_range" : null) : Invalid(value, null, streets.Cast<object>().ToArray());
    }

    private static bool? ContainsHouseNumber(Street street, long number)
    {
        var start = number % 2 == 0 ? street.EvenStart : street.OddStart;
        var end = number % 2 == 0 ? street.EvenEnd : street.OddEnd;
        if (!RangeNumber(start, out var first)) return null;
        if (Normalize(end ?? "") == "τελ") return number >= first;
        return RangeNumber(end, out var last) ? number >= first && number <= last : null;
    }

    private static bool PositiveInteger(object value, out long number) => value switch
    {
        byte n => Set(n, out number), short n when n > 0 => Set(n, out number), int n when n > 0 => Set(n, out number), long n when n > 0 => Set(n, out number),
        string text when long.TryParse(text, NumberStyles.None, CultureInfo.InvariantCulture, out var n) && n > 0 => Set(n, out number),
        _ => Set(0, out number, false),
    };
    private static bool Set(long value, out long output, bool valid = true) { output = value; return valid; }
    private static bool RangeNumber(string? value, out long number) => long.TryParse(RangeNumberPattern.Match(value ?? "").Groups[1].Value, out number);
    private static string Normalize(string value) => string.Concat(value.Normalize(NormalizationForm.FormD).ToLowerInvariant().Where(character => CharUnicodeInfo.GetUnicodeCategory(character) != UnicodeCategory.NonSpacingMark && char.IsLetterOrDigit(character)));
    private static void ValidateLimit(int? limit) { if (limit is <= 0) throw new ArgumentOutOfRangeException(nameof(limit), "Limit must be a positive integer."); }
    private static Entity Entity(SqliteDataReader row, bool officialCode) => new(row.GetInt64(0), row.GetString(1), officialCode && !row.IsDBNull(2) ? row.GetString(2) : null);
    private static Entity? Named(Entity? entity) => entity is null ? null : entity with { OfficialCode = null };
    private static Entity? Coded(Entity? entity) => entity;
    private static Street Street(SqliteDataReader row) => new(row.GetInt64(0), row.GetString(1), row.GetString(2), Text(row, 3), Text(row, 4), Text(row, 5), Text(row, 6));
    private static PostcodeResult Location(SqliteDataReader row) => new(row.GetString(0), Number(row, 1), Number(row, 2), Text(row, 3), Integer(row, 4), Integer(row, 5), Integer(row, 6));
    private static string? Text(SqliteDataReader row, int index) => row.IsDBNull(index) ? null : row.GetString(index);
    private static double? Number(SqliteDataReader row, int index) => row.IsDBNull(index) ? null : row.GetDouble(index);
    private static long? Integer(SqliteDataReader row, int index) => row.IsDBNull(index) ? null : row.GetInt64(index);
    private static ValidationResult Valid(object input, IReadOnlyList<object> matches, string? reason = null) => new(ValidationStatus.Valid, input, matches, reason);
    private static ValidationResult Invalid(object input, string? reason, IReadOnlyList<object>? matches = null) => new(ValidationStatus.Invalid, input, matches, reason);
    private static ValidationResult NotEvaluated(object input, string reason) => new(ValidationStatus.NotEvaluated, input, null, reason);
    private long? ScalarLong(string sql, long id) => One(sql, c => c.Parameters.AddWithValue("$id", id), row => row.GetInt64(0));
    private T? One<T>(string sql, Action<SqliteCommand> parameters, Func<SqliteDataReader, T> map) where T : class { using var command = Command(sql, parameters); using var reader = command.ExecuteReader(); return reader.Read() ? map(reader) : null; }
    private long? One(string sql, Action<SqliteCommand> parameters, Func<SqliteDataReader, long> map) { using var command = Command(sql, parameters); using var reader = command.ExecuteReader(); return reader.Read() ? map(reader) : null; }
    private IReadOnlyList<T> All<T>(string sql, Action<SqliteCommand> parameters, Func<SqliteDataReader, T> map) { using var command = Command(sql, parameters); using var reader = command.ExecuteReader(); var result = new List<T>(); while (reader.Read()) result.Add(map(reader)); return result; }
    private SqliteCommand Command(string sql, Action<SqliteCommand> parameters) { var command = Connection.CreateCommand(); command.CommandText = sql; parameters(command); return command; }
    private SqliteConnection Connection => _connection ?? throw new ObjectDisposedException(nameof(PostalCodeClient));
}

File: java/src/main/java/io/github/rallisf1/greekpostalcode/PostalCodeClient.java
package io.github.rallisf1.greekpostalcode;

import static io.github.rallisf1.greekpostalcode.Models.*;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.*;
import java.text.Normalizer;
import java.util.*;
import java.util.regex.Pattern;

/** Offline, read-only access to the Greek Postal Code DB bundled in this JAR. */
public final class PostalCodeClient implements AutoCloseable {
    private static final Pattern POSTCODE = Pattern.compile("\\d{5}");
    private static final Pattern RANGE_NUMBER = Pattern.compile("^\\s*(\\d+)");
    private Connection connection;
    private final Path databasePath;

    private PostalCodeClient(Connection connection, Path databasePath) { this.connection = connection; this.databasePath = databasePath; }

    /** Opens a temporary read-only copy of the SQLite database embedded in this JAR. */
    public static PostalCodeClient create() {
        Path path = null;
        try (InputStream input = PostalCodeClient.class.getResourceAsStream("/library.sqlite")) {
            if (input == null) throw new IllegalStateException("The bundled library.sqlite resource is missing.");
            path = Files.createTempFile("greek-postal-code-db-", ".sqlite");
            Files.copy(input, path, java.nio.file.StandardCopyOption.REPLACE_EXISTING);
            Connection connection = DriverManager.getConnection("jdbc:sqlite:file:" + path.toAbsolutePath() + "?mode=ro");
            try (Statement statement = connection.createStatement()) { statement.execute("PRAGMA query_only = ON"); }
            return new PostalCodeClient(connection, path);
        } catch (SQLException | IOException exception) {
            if (path != null) try { Files.deleteIfExists(path); } catch (IOException ignored) { }
            throw new PostalCodeException("Could not open the bundled SQLite database.", exception);
        }
    }

    @Override public void close() {
        if (connection == null) return;
        try { connection.close(); } catch (SQLException exception) { throw new PostalCodeException("Could not close the SQLite database.", exception); }
        finally { connection = null; try { Files.deleteIfExists(databasePath); } catch (IOException ignored) { } }
    }

    public List<Entity> listRegions(ListOptions options) { return list("regions", null, null, defaults(options), false); }
    public List<Entity> listRegions() { return listRegions(new ListOptions()); }
    public List<Entity> listRegionalUnits(RegionalUnitOptions options) { options = options == null ? new RegionalUnitOptions() : options; return list("regional_units", "region_id", options.regionId(), options.listOptions(), options.includeOfficialCode()); }
    public List<Entity> listMunicipalities(MunicipalityOptions options) { options = options == null ? new MunicipalityOptions() : options; return list("municipalities", "regional_unit_id", options.regionalUnitId(), options.listOptions(), options.includeOfficialCode()); }
    public List<Entity> listMunicipalUnits(MunicipalityChildOptions options) { options = options == null ? new MunicipalityChildOptions() : options; return list("municipal_units", "municipality_id", options.municipalityId(), options.listOptions(), options.includeOfficialCode()); }
    public List<Entity> listCommunities(MunicipalityChildOptions options) { options = options == null ? new MunicipalityChildOptions() : options; return list("communities", "municipality_id", options.municipalityId(), options.listOptions(), options.includeOfficialCode()); }

    public List<Entity> searchRegions(String query, ListOptions options) { return search("regions", null, query, null, defaults(options), false); }
    public List<Entity> searchRegions(String query) { return searchRegions(query, new ListOptions()); }
    public List<Entity> searchRegionalUnits(String query, RegionalUnitOptions options) { options = options == null ? new RegionalUnitOptions() : options; return search("regional_units", "region_id", query, options.regionId(), options.listOptions(), options.includeOfficialCode()); }
    public List<Entity> searchMunicipalities(String query, MunicipalityOptions options) { options = options == null ? new MunicipalityOptions() : options; return search("municipalities", "regional_unit_id", query, options.regionalUnitId(), options.listOptions(), options.includeOfficialCode()); }
    public List<Entity> searchMunicipalUnits(String query, MunicipalityChildOptions options) { options = options == null ? new MunicipalityChildOptions() : options; return search("municipal_units", "municipality_id", query, options.municipalityId(), options.listOptions(), options.includeOfficialCode()); }
    public List<Entity> searchCommunities(String query, MunicipalityChildOptions options) { options = options == null ? new MunicipalityChildOptions() : options; return search("communities", "municipality_id", query, options.municipalityId(), options.listOptions(), options.includeOfficialCode()); }

    public PostcodeResult getPostcode(String postcode, PostcodeOptions options) {
        if (postcode == null || !POSTCODE.matcher(postcode).matches()) return null;
        options = options == null ? new PostcodeOptions() : options;
        PostcodeResult location = one("SELECT postcode, latitude, longitude, local_area, municipal_unit_id, community_id, municipality_id FROM locations WHERE postcode = ?", statement -> statement.setString(1, postcode), this::location);
        if (location == null) return null;
        return new PostcodeResult(location.postcode(), location.latitude(), location.longitude(), location.localArea(), location.municipalUnitId(), location.communityId(), location.municipalityId(), options.includeHierarchy() ? postcodeHierarchy(location) : null, options.includeStreets() ? streets(postcode) : null);
    }
    public PostcodeResult getPostcode(String postcode) { return getPostcode(postcode, new PostcodeOptions()); }

    public AddressValidation validateAddress(AddressInput input) {
        Objects.requireNonNull(input, "input");
        PostcodeResult postcode = getPostcode(input.postcode(), new PostcodeOptions(true, input.street() != null || input.houseNumber() != null));
        ValidationResult postcodeResult = input.postcode() == null || !POSTCODE.matcher(input.postcode()).matches() ? invalid(input.postcode(), "postcode_must_be_exactly_five_digits", null) : postcode == null ? invalid(input.postcode(), "postcode_not_found", null) : valid(input.postcode(), List.of(postcode), null);
        if (postcode == null) return new AddressValidation(postcodeResult, ne(input.street(), "postcode_not_found"), ne(input.houseNumber(), "postcode_not_found"), ne(input.municipality(), "postcode_not_found"), ne(input.municipalUnit(), "postcode_not_found"), ne(input.community(), "postcode_not_found"), ne(input.regionalUnit(), "postcode_not_found"), ne(input.region(), "postcode_not_found"));
        EntityHierarchy hierarchy = postcode.hierarchy();
        ValidationResult street = input.street() == null ? null : validateStreet(input.street(), postcode.streets());
        return new AddressValidation(postcodeResult, street, input.houseNumber() == null ? null : validateHouseNumber(input.houseNumber(), street),
            input.municipality() == null ? null : validateReference("municipalities", input.municipality(), hierarchy.municipality()),
            input.municipalUnit() == null ? null : validateReference("municipal_units", input.municipalUnit(), hierarchy.municipalUnit()),
            input.community() == null ? null : validateReference("communities", input.community(), hierarchy.community()),
            input.regionalUnit() == null ? null : validateReference("regional_units", input.regionalUnit(), hierarchy.regionalUnit()),
            input.region() == null ? null : validateReference("regions", input.region(), hierarchy.region()));
    }

    private List<Entity> list(String table, String parentColumn, Long parentId, ListOptions options, boolean officialCode) {
        limit(options.limit());
        String sql = "SELECT id, name" + (officialCode ? ", official_code" : "") + " FROM " + table + (parentColumn != null && parentId != null ? " WHERE " + parentColumn + " = ?" : "") + " ORDER BY name, id";
        List<Entity> result = all(sql, statement -> { if (parentColumn != null && parentId != null) statement.setLong(1, parentId); }, row -> entity(row, officialCode));
        if (options.includeHierarchy()) result = result.stream().map(value -> new Entity(value.id(), value.name(), value.officialCode(), entityHierarchy(table, value.id()))).toList();
        return options.limit() == null ? result : result.subList(0, Math.min(options.limit(), result.size()));
    }
    private List<Entity> search(String table, String parentColumn, String query, Long parentId, ListOptions options, boolean officialCode) {
        String normalized = normalize(query);
        if (normalized.isEmpty()) return List.of();
        List<Entity> matches = list(table, parentColumn, parentId, new ListOptions(null, options.includeHierarchy(), options.includeOfficialCode()), officialCode).stream().filter(value -> normalize(value.name()).startsWith(normalized)).toList();
        return options.limit() == null ? matches : matches.subList(0, Math.min(options.limit(), matches.size()));
    }
    private EntityHierarchy entityHierarchy(String table, long id) {
        return switch (table) {
            case "regions" -> new EntityHierarchy(named(one("SELECT da.id, da.name FROM regions r JOIN decentralized_administrations da ON da.id = r.decentralized_administration_id WHERE r.id = ?", s -> s.setLong(1, id), row -> entity(row, false))), null, null, null, null, null);
            case "regional_units" -> regionalUnitHierarchy(id);
            case "municipalities" -> municipalityHierarchy(id);
            default -> municipalityChildHierarchy(one("SELECT municipality_id FROM " + table + " WHERE id = ?", s -> s.setLong(1, id), row -> row.getLong(1)));
        };
    }
    private EntityHierarchy postcodeHierarchy(PostcodeResult location) {
        Entity unit = location.municipalUnitId() == null ? null : coded(one("SELECT id, name, official_code FROM municipal_units WHERE id = ?", s -> s.setLong(1, location.municipalUnitId()), row -> entity(row, true)));
        Entity community = location.communityId() == null ? null : coded(one("SELECT id, name, official_code FROM communities WHERE id = ?", s -> s.setLong(1, location.communityId()), row -> entity(row, true)));
        Entity municipality = location.municipalityId() == null ? null : coded(one("SELECT id, name, official_code FROM municipalities WHERE id = ?", s -> s.setLong(1, location.municipalityId()), row -> entity(row, true)));
        EntityHierarchy parent = municipality == null ? emptyHierarchy() : municipalityHierarchy(municipality.id());
        return new EntityHierarchy(parent.decentralizedAdministration(), parent.region(), parent.regionalUnit(), municipality, unit, community);
    }
    private EntityHierarchy regionalUnitHierarchy(long id) {
        Entity region = named(one("SELECT r.id, r.name FROM regional_units ru JOIN regions r ON r.id = ru.region_id WHERE ru.id = ?", s -> s.setLong(1, id), row -> entity(row, false)));
        Entity administration = named(one("SELECT da.id, da.name FROM regional_units ru JOIN regions r ON r.id = ru.region_id JOIN decentralized_administrations da ON da.id = r.decentralized_administration_id WHERE ru.id = ?", s -> s.setLong(1, id), row -> entity(row, false)));
        return new EntityHierarchy(administration, region, null, null, null, null);
    }
    private EntityHierarchy municipalityHierarchy(long id) {
        Entity unit = coded(one("SELECT ru.id, ru.name, ru.official_code FROM municipalities m JOIN regional_units ru ON ru.id = m.regional_unit_id WHERE m.id = ?", s -> s.setLong(1, id), row -> entity(row, true)));
        EntityHierarchy parent = unit == null ? emptyHierarchy() : regionalUnitHierarchy(unit.id());
        return new EntityHierarchy(parent.decentralizedAdministration(), parent.region(), unit, null, null, null);
    }
    private EntityHierarchy municipalityChildHierarchy(Long id) {
        if (id == null) return emptyHierarchy();
        Entity municipality = coded(one("SELECT id, name, official_code FROM municipalities WHERE id = ?", s -> s.setLong(1, id), row -> entity(row, true)));
        EntityHierarchy parent = municipality == null ? emptyHierarchy() : municipalityHierarchy(municipality.id());
        return new EntityHierarchy(parent.decentralizedAdministration(), parent.region(), parent.regionalUnit(), municipality, null, null);
    }
    private List<Street> streets(String postcode) { return all("SELECT id, postcode, name, odd_start, odd_end, even_start, even_end FROM streets WHERE postcode = ? ORDER BY name, id", s -> s.setString(1, postcode), this::street); }
    private ValidationResult validateReference(String table, LocalityReference reference, Entity linked) {
        boolean officialCode = !table.equals("regions");
        List<Entity> matches;
        if (reference instanceof LocalityId id) matches = all("SELECT id, name" + (officialCode ? ", official_code" : "") + " FROM " + table + " WHERE id = ?", s -> s.setLong(1, id.id()), row -> entity(row, officialCode));
        else if (reference instanceof LocalityName name && !normalize(name.name()).isEmpty()) matches = list(table, null, null, new ListOptions(), officialCode).stream().filter(value -> normalize(value.name()).startsWith(normalize(name.name()))).toList();
        else return invalid(reference, "reference_must_not_be_empty", List.of());
        if (linked == null) return invalid(reference, "postcode_has_no_linked_entity", matches);
        return matches.stream().anyMatch(value -> value.id() == linked.id()) ? valid(reference, matches, null) : invalid(reference, null, matches);
    }
    private static ValidationResult validateStreet(String input, List<Street> streets) { List<Street> matches = streets.stream().filter(value -> normalize(value.name()).equals(normalize(input))).toList(); return matches.isEmpty() ? invalid(input, "street_not_found_for_postcode", matches) : valid(input, matches, null); }
    private static ValidationResult validateHouseNumber(Object input, ValidationResult street) {
        if (street == null || street.status() != ValidationStatus.VALID) return ne(input, "street_is_required_and_must_be_valid");
        Long number = positiveInteger(input); if (number == null) return invalid(input, "house_number_must_be_a_positive_integer", null);
        @SuppressWarnings("unchecked") List<Street> streets = (List<Street>) street.matches();
        List<Boolean> checks = streets.stream().map(value -> contains(value, number)).filter(Objects::nonNull).toList();
        return checks.isEmpty() || checks.contains(true) ? valid(input, streets, checks.isEmpty() ? "street_has_no_usable_range" : null) : invalid(input, null, streets);
    }
    private static Boolean contains(Street street, long number) { String start = number % 2 == 0 ? street.evenStart() : street.oddStart(); String end = number % 2 == 0 ? street.evenEnd() : street.oddEnd(); Long first = rangeNumber(start); if (first == null) return null; if (normalize(end == null ? "" : end).equals("τελ")) return number >= first; Long last = rangeNumber(end); return last == null ? null : number >= first && number <= last; }
    private static Long positiveInteger(Object value) { if (value instanceof Number number && number.longValue() > 0 && number.doubleValue() == number.longValue()) return number.longValue(); if (value instanceof String text && text.matches("\\d+") && Long.parseLong(text) > 0) return Long.parseLong(text); return null; }
    private static Long rangeNumber(String value) { var match = RANGE_NUMBER.matcher(value == null ? "" : value); return match.find() ? Long.valueOf(match.group(1)) : null; }
    private static String normalize(String value) { if (value == null) return ""; String normalized = Normalizer.normalize(value, Normalizer.Form.NFD).toLowerCase(Locale.ROOT); StringBuilder result = new StringBuilder(); normalized.codePoints().filter(c -> Character.getType(c) != Character.NON_SPACING_MARK && Character.isLetterOrDigit(c)).forEach(result::appendCodePoint); return result.toString(); }
    private static void limit(Integer value) { if (value != null && value <= 0) throw new IllegalArgumentException("limit must be a positive integer"); }
    private static ListOptions defaults(ListOptions options) { return options == null ? new ListOptions() : options; }
    private static EntityHierarchy emptyHierarchy() { return new EntityHierarchy(null, null, null, null, null, null); }
    private static Entity named(Entity entity) { return entity == null ? null : new Entity(entity.id(), entity.name(), null, null); }
    private static Entity coded(Entity entity) { return entity; }
    private static ValidationResult valid(Object input, List<?> matches, String reason) { return new ValidationResult(ValidationStatus.VALID, input, matches, reason); }
    private static ValidationResult invalid(Object input, String reason, List<?> matches) { return new ValidationResult(ValidationStatus.INVALID, input, matches, reason); }
    private static ValidationResult ne(Object input, String reason) { return input == null ? null : new ValidationResult(ValidationStatus.NOT_EVALUATED, input, null, reason); }
    private Entity entity(ResultSet row, boolean officialCode) throws SQLException { return new Entity(row.getLong(1), row.getString(2), officialCode ? row.getString(3) : null, null); }
    private Street street(ResultSet row) throws SQLException { return new Street(row.getLong(1), row.getString(2), row.getString(3), row.getString(4), row.getString(5), row.getString(6), row.getString(7)); }
    private PostcodeResult location(ResultSet row) throws SQLException { return new PostcodeResult(row.getString(1), nullableDouble(row, 2), nullableDouble(row, 3), row.getString(4), nullableLong(row, 5), nullableLong(row, 6), nullableLong(row, 7), null, null); }
    private static Long nullableLong(ResultSet row, int column) throws SQLException { long value = row.getLong(column); return row.wasNull() ? null : value; }
    private static Double nullableDouble(ResultSet row, int column) throws SQLException { double value = row.getDouble(column); return row.wasNull() ? null : value; }
    private <T> T one(String sql, Binder binder, Mapper<T> mapper) { List<T> values = all(sql, binder, mapper); return values.isEmpty() ? null : values.getFirst(); }
    private <T> List<T> all(String sql, Binder binder, Mapper<T> mapper) { try (PreparedStatement statement = connection().prepareStatement(sql)) { binder.bind(statement); try (ResultSet rows = statement.executeQuery()) { List<T> result = new ArrayList<>(); while (rows.next()) result.add(mapper.map(rows)); return result; } } catch (SQLException exception) { throw new PostalCodeException("SQLite query failed.", exception); } }
    private Connection connection() { if (connection == null) throw new IllegalStateException("PostalCodeClient is closed."); return connection; }
    @FunctionalInterface private interface Binder { void bind(PreparedStatement statement) throws SQLException; }
    @FunctionalInterface private interface Mapper<T> { T map(ResultSet row) throws SQLException; }
    public static final class PostalCodeException extends RuntimeException { public PostalCodeException(String message, Throwable cause) { super(message, cause); } }
}
